apps/hdl_graph_slam_nodelet.cpp

```diff
@@ -802,6 +808,103 @@ class HdlGraphSlamNodelet : public nodelet::Nodelet {
+   * @param req
+   * @param res
+   * @return
+   */
+  bool load_service(hdl_graph_slam::LoadGraphRequest& req, hdl_graph_slam::LoadGraphResponse& res) {
+    std::lock_guard<std::mutex> lock(main_thread_mutex);
+
+    std::string directory = req.path;
+
+    std::cout << "loading data from:" << directory << std::endl;
+
+    // Load graph.
+    graph_slam->load(directory + "/graph.g2o");
+    
+    // Iterate over the items in this directory and count how many sub directories there are. 
+    // This will give an upper limit on how many keyframe indexes we can expect to find.
+    boost::filesystem::directory_iterator begin(directory), end;
+    int max_directory_count = std::count_if(begin, end,
+        [](const boost::filesystem::directory_entry & d) {
+            return boost::filesystem::is_directory(d.path()); // only return true if a direcotry
+    });
+
+    // Load keyframes by looping through key frame indexes that we expect to see.
+    for(int i = 0; i < max_directory_count; i++) {
+      std::stringstream sst;
+      sst << boost::format("%s/%06d") % directory % i;
+      std::string key_frame_directory = sst.str();
+
+      // If key_frame_directory doesnt exist, then we have run out so lets stop looking.
+      if(!boost::filesystem::is_directory(key_frame_directory)) {
+        break;
+      }
+
+      KeyFrame::Ptr keyframe(new KeyFrame(key_frame_directory, graph_slam->graph.get()));
+      keyframes.push_back(keyframe);
+    }
+    std::cout << "loaded " << keyframes.size() << " keyframes" <<std::endl;
+    
+    // Load special nodes.
+    std::ifstream ifs(directory + "/special_nodes.csv");
+    if(!ifs) {
+      return false;
+    }
+    while(!ifs.eof()) {
+      std::string token;
+      ifs >> token;
+      if(token == "anchor_node") {
+        int id = 0;
+        ifs >> id;
+        anchor_node = static_cast<g2o::VertexSE3*>(graph_slam->graph->vertex(id));
+      } else if(token == "anchor_edge") {
+        int id = 0;
+        ifs >> id; 
+        // We have no way of directly pulling the edge based on the edge ID that we have just read in.
+        // Fortunatly anchor edges are always attached to the anchor node so we can iterate over 
+        // the edges that listed against the node untill we find the one that matches our ID.
+        if(anchor_node){
+          auto edges = anchor_node->edges();
+
+          for(auto e : edges) {
+            int edgeID =  e->id();
+            if (edgeID == id){
+              anchor_edge = static_cast<g2o::EdgeSE3*>(e);
+
+              break;
+            }
+          } 
+        }
+      } else if(token == "floor_node") {
+        int id = 0;
+        ifs >> id;
+        floor_plane_node = static_cast<g2o::VertexPlane*>(graph_slam->graph->vertex(id));
+      }
+    }
+    std::cout << "loaded special nodes - anchor_node: "<< anchor_node->id() << " anchor_edge: "<< anchor_edge->id() << " floor_node: "<< floor_plane_node->id() << std::endl;
```

Comment:
there is a minor issue in this line. If either `anchor_node`, `anchor_edge` or `floor_plane_node` is a null pointer, then we cannot print `->id()` for them.
I'd suggest only printing when they're not null pointers. I think that only `anchor_node` needs to be non-null to be able to load the graph successfully